exchanges/stream/buffer/buffer.go

```diff
@@ -42,246 +78,350 @@ func (w *Orderbook) Update(u *Update) error {
+func insertBid(updt orderbook.Item, book *[]orderbook.Item) {
+	for target := range *book {
+		if updt.Price > []orderbook.Item(*book)[target].Price {
+			insertItem(updt, book, target)
+			return
+		}
 	}
+	*book = append(*book, updt)
+}
 
-	w.m.Lock()
-	defer w.m.Unlock()
-	if w.ob == nil {
-		w.ob = make(map[currency.Pair]map[asset.Item]*orderbook.Base)
+// insertUpdatesBid inserts on **correctly aligned** book at price level
+func insertUpdatesBid(updt []orderbook.Item, book *[]orderbook.Item) {
+updates:
+	for x := range updt {
+		for target := range *book {
+			if updt[x].Price > []orderbook.Item(*book)[target].Price {
+				insertItem(updt[x], book, target)
+				continue updates
+			}
+		}
+		*book = append(*book, updt[x])
 	}
-	if w.ob[newOrderbook.Pair] == nil {
-		w.ob[newOrderbook.Pair] = make(map[asset.Item]*orderbook.Base)
+}
+
+// insertUpdatesBid inserts on **correctly aligned** book at price level
+func insertUpdatesAsk(updt []orderbook.Item, book *[]orderbook.Item) {
+updates:
+	for x := range updt {
+		for target := range *book {
+			if updt[x].Price < []orderbook.Item(*book)[target].Price {
+				insertItem(updt[x], book, target)
+				continue updates
+			}
+		}
+		*book = append(*book, updt[x])
 	}
+}
+
+// insertItem inserts item in slice by target element this is an optimization
+// to reduce the need for sorting algorithms
+func insertItem(update orderbook.Item, book *[]orderbook.Item, target int) {
+	// TODO: extend slice by incoming update length before this gets hit
+	*book = append(*book, orderbook.Item{})
+	copy([]orderbook.Item(*book)[target+1:], []orderbook.Item(*book)[target:])
+	[]orderbook.Item(*book)[target] = update
+}
 
-	w.ob[newOrderbook.Pair][newOrderbook.AssetType] = newOrderbook
-	err := newOrderbook.Process()
+// LoadSnapshot loads initial snapshot of ob data from websocket
+func (w *Orderbook) LoadSnapshot(book *orderbook.Base) error {
+	w.m.Lock()
+	defer w.m.Unlock()
+
+	err := book.Process()
 	if err != nil {
 		return err
 	}
 
-	w.dataHandler <- newOrderbook
+	m1, ok := w.ob[book.Pair.Base]
+	if !ok {
+		m1 = make(map[currency.Code]map[asset.Item]*orderbook.Base)
+		w.ob[book.Pair.Base] = m1
+	}
+	m2, ok := m1[book.Pair.Quote]
+	if !ok {
+		m2 = make(map[asset.Item]*orderbook.Base)
+		m1[book.Pair.Quote] = m2
+	}
+	m3, ok := m2[book.AssetType]
+	if !ok {
+		m3 = book
+		m2[book.AssetType] = m3
+	} else {
+		m3.Bids = book.Bids
+		m3.Asks = book.Asks
+	}
+
+	w.dataHandler <- book
 	return nil
 }
 
-// GetOrderbook use sparingly. Modifying anything here will ruin hash
-// calculation and cause problems
+// GetOrderbook returns orderbook stored in current buffer
 func (w *Orderbook) GetOrderbook(p currency.Pair, a asset.Item) *orderbook.Base {
 	w.m.Lock()
-	ob := w.ob[p][a]
-	w.m.Unlock()
-	return ob
+	defer w.m.Unlock()
+	ptr, ok := w.ob[p.Base][p.Quote][a]
+	if !ok {
+		return nil
+	}
+	cpy := *ptr
+	cpy.Asks = append(ptr.Asks[:0:0], ptr.Asks...) // nolint:gocritic
```

Comment:
Please add additional comments on why you're using nolint

Reply:
Just a shorthand to not use make and a copy. Will change.